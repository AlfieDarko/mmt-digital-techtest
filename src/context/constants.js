export const CLEAR_BASKET = "CLEAR_BASKET"
export const CHANGE_QUANTITY = "CHANGE_QUANTITY"