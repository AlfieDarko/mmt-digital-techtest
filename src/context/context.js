import { createContext, useContext } from 'react'

export const ProductBasketStateCtx = createContext()
export const ProductBasketDispatchCtx = createContext()


export const useProductBasketState = () => {
  const context = useContext(ProductBasketStateCtx);
  if (context === undefined) {
    throw new Error(
      "useProductBasketState() must be used within a ProductBasket Provider. Check you are using it correctly."
    );
  }
  return context;
}

export const useProductBasketDispatch = () => {
  const context = useContext(ProductBasketDispatchCtx);
  if (context === undefined) {
    throw new Error(
      "useProductBasketDispatch() must be used within a ProductBasket Provider. Check you are using it correctly."
    );
  }
  return context;
}

export const useProductBasket = () => {
  return [useProductBasketState(), useProductBasketDispatch()]
}
