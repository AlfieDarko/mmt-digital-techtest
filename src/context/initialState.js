export const initBasketState = {
  products: {
    byId: {
      "drinkid1": {
        id: 'drinkid1',
        title: "Mountain Dew",
        description: "Sugar, Citric acid with a splash of Hydrofluoric Acid & water",
        price: 109,
        quantity: 0,
        imageSRC: "https://digitalcontent.api.tesco.com/v2/media/ghs/5cc66364-c063-48f1-b9e0-8d06bcb5b857/snapshotimagehandler_1451681747.jpeg"
      },
      "drinkid2": {
        id: 'drinkid2',
        title: "Desperados",
        description: "Tequila flavoured, Mexican inspired, French created, Croatian produced beer.",
        price: 599,
        quantity: 0,
        imageSRC: "https://www.thedesignateddriver.co.uk/wp-content/uploads/2019/08/Desperados-600x600.jpg"
      },
      "drinkid3": {
        id: 'drinkid3',
        title: "Jack Daniels",
        description: "Moonshine but rebranded.",
        price: 3499,
        quantity: 0,
        imageSRC: "https://www.latramuntana.store/545-large_default/jack-daniels-amb-balanci-3-l.jpg"
      }
    },
    allIds: ['drinkid1', 'drinkid2', 'drinkid3']
  }
}