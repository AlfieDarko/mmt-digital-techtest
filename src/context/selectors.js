export const selectProductById = (state, id) => {
  return state?.products.byId[id]
}

export const getNewProductQuantity = (state, id, changeInQuantity) => {
  const product = selectProductById(state, id)
  if (product.quantity === 0 && changeInQuantity < 0) {
    throw new Error('Cannot reduce quantity below 0. Please try again')
  }
  return product.quantity + changeInQuantity
}

export const getProducts = (state) => {
  return state.products.byId
}