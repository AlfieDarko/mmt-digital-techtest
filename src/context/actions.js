import { CLEAR_BASKET, CHANGE_QUANTITY } from './constants'

export const clearBasket = (dispatch) => {
  dispatch({ type: CLEAR_BASKET })
}

export const changeProductQuantity = (id, quantity, dispatch) => {
  dispatch({ type: CHANGE_QUANTITY, payload: { id, quantity } })
}