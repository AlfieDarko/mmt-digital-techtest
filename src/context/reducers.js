import React from 'react'
import { initBasketState } from './initialState'

export const productBasketReducer = (state, action) => {

  switch (action.type) {
    case "CHANGE_QUANTITY": {
      const { payload: { id, quantity } } = action;
      return {
        ...state, products: {
          ...state.products, byId: {
            ...state.products.byId, [id]: {
              ...state.products.byId[id], quantity
            }
          }
        }
      }
    }

    case "CLEAR_BASKET": {
      return {
        ...initBasketState
      }
    }

    default: {
      return state
    }
  }
}