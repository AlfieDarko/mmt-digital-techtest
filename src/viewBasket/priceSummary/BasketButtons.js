import React from 'react'
import { Flex, Text, Button } from '@chakra-ui/core'
import { useProductBasket } from '../../context/context'
import { CLEAR_BTN_TEXT, CHECKOUT_BTN_TEXT } from '../constants'

const BasketButtons = () => {
  const [, dispatch] = useProductBasket()
  const clearOnClickHandler = () => {
    dispatch({ type: 'CLEAR_BASKET' })
  }

  return (
    <>
      <Flex mt="8" py="8" pr="4" alignItems="center" justifyContent="space-between" bg="#F8F8F7">
        <Button rol ml={8} variant="outline" onClick={clearOnClickHandler} data-testid="clear-btn">
          {CLEAR_BTN_TEXT}
        </Button>
        <Button color="white" bg="#44b8f3" borderRadius="5px" p={5} data-testid="checkout-btn">
          <Text fontWeight="bold" color="#1b212c">{CHECKOUT_BTN_TEXT}</Text>
        </Button>
      </Flex>
    </>
  )
}

export default BasketButtons
