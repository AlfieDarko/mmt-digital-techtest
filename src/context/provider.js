import React from 'react'
import { ProductBasketStateCtx, ProductBasketDispatchCtx } from './context'
import { productBasketReducer } from './reducers'
import { useReducer } from 'react'
import { initBasketState } from './initialState'

export const ProductBasketProvider = ({ children }) => {
  const [state, dispatch] = useReducer(productBasketReducer, initBasketState)

  return (
    <ProductBasketStateCtx.Provider value={state}>
      <ProductBasketDispatchCtx.Provider value={dispatch}>
        {children}
      </ProductBasketDispatchCtx.Provider>
    </ProductBasketStateCtx.Provider>
  )
}

