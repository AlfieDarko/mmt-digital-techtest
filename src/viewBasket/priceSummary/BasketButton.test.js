import React from "react";
import { render } from '@testing-library/react'
import { ThemeProvider } from '@chakra-ui/core'
import BasketButtons from './'
import { ProductBasketProvider } from '../../context/provider'

const customRender = (ui, {
  providerProps, ...renderOptions
}) => {
  return render(
    <ProductBasketProvider {...providerProps}>
      <ThemeProvider>
        {ui}
      </ThemeProvider>
    </ProductBasketProvider>, renderOptions
  )
}

describe('ProdutInformation', () => {
  const props = {
    imageSRC: "",
    title: "test product",
    description: "product description",
    price: 100,
    quantity: 2,
    id: 'beer1'
  }

  const providerProps = {
    state: {
      products: {
        byId: {
          "beer1": props
        }
      }
    }
  }

  it('Should match snapshots', () => {
    const { asFragment } = customRender(<BasketButtons />, { providerProps });
    expect(asFragment()).toMatchSnapshot();
  });
});