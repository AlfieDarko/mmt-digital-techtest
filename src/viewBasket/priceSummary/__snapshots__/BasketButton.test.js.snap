// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProdutInformation Should match snapshots 1`] = `
<DocumentFragment>
  <section
    class="css-k008qs"
    data-testid="price-summary"
  >
    <div
      class="css-1bpae1b"
    >
      <h2
        class="css-1jrokaa"
      >
        Summary
      </h2>
      <div
        class="css-0"
      >
        <div
          class="css-13rlfrq"
          data-testid="total-amount"
        >
          <span
            class="css-15619qa"
          >
            Total:
          </span>
          <span
            class="css-f1b05"
            data-testid="total-price"
          >
            Â£0.00
          </span>
        </div>
        <div
          class="css-13rlfrq"
        >
          <span
            class="css-15619qa"
          >
            Shipping:
          </span>
          <span
            class="css-grr4rk"
          >
            free
          </span>
        </div>
      </div>
      <div
        class="css-cy9gus"
      >
        <button
          class="css-u0mtjy"
          data-testid="clear-btn"
          type="button"
        >
          Clear
        </button>
        <button
          class="css-1jweop2"
          data-testid="checkout-btn"
          type="button"
        >
          <p
            class="css-84qp33"
          >
            Check Out &gt;
          </p>
        </button>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
